### DISCUSSION TITLE
Running autokey service on i3wm

### DISCUSSION THREAD

**Discussion by stibinator** [autokey/autokey#991] [Q&A] [answered]
I'm using i3wm on Arch, and I can't get Autokey to work in the background. In my `~/config/i3/config` file if I use 
```
exec --no-startup-id autokey-gtk -c
```
the Autokey settings window opens at login, and my autokey phrases work, but only as long as the settings window is open. I can't work out how to run autokey as a service. I tried using systemctl, but got: 
```
➜  ~ systemctl enable autokey.service
[..]
Failed to enable unit: Unit autokey.service does not exist
```
If I run autokey-gtk (or autokey-qt) manually from a command line, my phrases work (though it complains about BadWindow errors), but only as long as I leave the process going in a command line.

**Comment by josephj11** [comment]
Welcome t te AutoKey community, @stibinator !

I'm not sure about this, but here are some thoughts.

AutoKey is not setup as a service.

IDK about your startup command. I haven't seen it done that way before.
When you start AutoKey, go to Settings->Configure AutoKey->Genreral->Automatically Start AutoKey After login and make sure it's selected and that your preferred/installed front end is selected.

If you can't figure out the startup command (and we can't find someone who knows i3wm), you can always start it from a terminal with

```
autokey-gtk &
```

The `&` will keep it running in the background without tying up the terminal, but it will still output some messages when starting or if you tell it to quit.

If you know where your system looks for startup scripts after login, you can put an executable shell script (with the appropriate shebang) there with the above command in it and then it won't send any "extra" messages to  a terminal you're actually using interactively. I know how to find those for KDE and Gnome, but not for i3wm. I will post a question on our forums.

The bad window messages are hard to pin down, but they usually don't cause problems. AFAIK, they are usually caused by defining hotkeys that are already being used by your desktop or window manager...

**Comment by jack99999** [comment]
i start autokey like this in my i3 config:

`exec /usr/bin/autokey-qt`

i've never tried to run it as a service

jack

**Comment by kreezxil** [comment]
do what @jack99999 did, when I tried i3wm, that's what I sued. Also don't use gtk, for some reason it has issues. I'm now on Linux Mint 22 for my main rig. GTK is what it wanted to install. but i switched it to qt so that autokey will behave.

IMO, qt is better than gtk. something to do with how windows are created/managed is my guess. I've not put too much effort into knowing exactly why.

**Comment by stibinator** [comment, marked as answer]
I put `exec --no-startup-id autokey-qt &` in my i3wm config, which starts autokey at login, but it behaves weirdly. If I have any terminal window open, I get messages from autokey appearing. I've given up on it for i3wm and am using espanso instead, thanks.

**Comment by josephj11** [comment]
@stibinator You shouldn't need  the ampersand  in a startup script and probably not  the flag either.

It's  not clear how that instance of AutoKey could be connected to another terminal's window/prcess. That's genuinely weird.
However, the messages should be easy enough to supress. Try

`exec  autokey-qt &> /dev/null &`

That should do it.

In your first post, you included the -c flag. Usually you woudn't want that in a startup sequence. Most of the time you're just using AutoKey, not coonfgurig it, so you don't need it  to display its configuration window by default.

**Comment by Elliria** [comment]
This is interesting. Since it's a script that's being auto-run on boot, it seems as though it would be launched from a non-interactive session rather than an open, interactive terminal window, so it shouldn't affect any terminal windows that are opened afterwards.

Would you consider using a clone of AutoKey instead of installing it, @stibinator? We have [the steps for that in the AutoKey wiki](https://github.com/autokey/autokey/wiki/Running-Tests#trying-out-a-clone-of-autokey) and it's actually the only way I run AutoKey in my virtual machine, although I also have AutoKey installed in there.

**Comment by Elliria** [comment]
I thought I'd add three links that might prove useful in debugging the behavior of the background job mentioned above:
* https://stackoverflow.com/questions/40079378/how-to-create-a-job-in-shell-script
* https://stackoverflow.com/questions/10408816/how-do-i-use-the-nohup-command-without-getting-nohup-out/10408906#10408906
* https://www.reddit.com/r/linuxquestions/comments/ypkcj6/terminal_how_do_i_run_a_command_send_it_to_the/
